File: 02-week-2--recursion/05-day-5--recursive-selection-sort/solutions/benchmark/benchmark.js
// Paste your iterative and recursive solutions in this file
// And then calculate their average run times to compare them.
function selectionSort(arr) {
  const sorted = [];

  while (arr.length > 0) {
    const min = Math.min(...arr);
    const idx = arr.indexOf(min);

    sorted.push(min);
    arr.splice(idx, 1);
  }

  return sorted;
}

function selectionSortRecursive(arr) {
  if (arr.length === 0) {
    return [];
  }

  const min = Math.min(...arr);
  const idx = arr.indexOf(min);
  arr.splice(idx, 1);

  const result = selectionSortRecursive(arr);
  result.unshift(min);
  return result;
}


function benchmark(callback) {
  const startTime = Date.now();

  for (let i = 0; i < 1000; ++i) {
    callback();
  }

  return (Date.now() - startTime) / 1000;
}

console.log('Iterative:', benchmark(() => selectionSort([2, 3, 4, 1, 4, 56, -2, 20])));
console.log('Recursive:', benchmark(() => selectionSortRecursive([2, 3, 4, 1, 4, 56, -2, 20])));

File: 02-week-2--recursion/05-day-5--recursive-selection-sort/solutions/benchmark/benchmark.rb
# Paste your iterative and recursive solutions in this file
# And then calculate their average run times to compare them.
def selection_sort(arr)
  sorted = []

  until arr.length == 0
    min = arr.min
    idx = arr.index(min)
    sorted << min
    arr.delete_at(idx)
  end

  sorted
end

def selection_sort_recursive(arr)
  return [] if arr.empty?

  min = arr.min
  idx = arr.index(min)
  arr.delete_at(idx)
  
  result = selection_sort_recursive(arr)
  result.unshift(min)
end

def benchmark
  start_time = Time.now

  1000.times do 
    yield
  end

  (Time.now - start_time) / 1000
end

puts "Iterative: #{benchmark { selection_sort([2, 3, 4, 1, 4, 56, -2, 20]) }}"
puts "Recursive: #{benchmark { selection_sort_recursive([2, 3, 4, 1, 4, 56, -2, 20]) }}"
